src/v/http/tests/request_builder_test.cc

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "http/request_builder.h"
+
+#include <gtest/gtest.h>
+
+namespace bh = boost::beast::http;
+
+TEST(request_builder, test_default_build_is_error) {
+    http::request_builder rb;
+    auto req = rb.build();
+    ASSERT_TRUE(!req.has_value());
+    ASSERT_TRUE(req.error() == rb.default_state);
+}
+
+TEST(request_builder, test_with_everything) {
+    http::request_builder rb;
+    auto req = rb.host("https://foobar.xyz")
+                 .target("san_mig")
+                 .with_bearer_auth("secret")
+                 .with_content_type("obsidian")
+                 .with_size_of(iobuf::from("a"))
+                 .header("one", "1")
+                 .headers({{"two", "2"}, {"three", "3"}})
+                 .method(boost::beast::http::verb::delete_)
+                 .query_param("here")
+                 .query_param_kv("name", "no")
+                 .query_params({"g", "b", "u"})
+                 .query_params_kv({{"trilogy", "of"}, {"containers", "four"}})
+                 .build();
+    ASSERT_TRUE(req.has_value());
+    const auto& built = req.value();
+    ASSERT_EQ(built.at(bh::field::host), "foobar.xyz");
```

Comment:
Generally in gtest, if the thing you're asserting isn't going to lead to UB or crash after, it's recommended to use `EXPECT`. So your `ASSERT_TRUE` above is good because otherwise `req.value` would fail. Although I think `at` fails if the key is not present? I'm not sure the types at play here, but maybe the advanced gtest matchers would be helpful: http://google.github.io/googletest/reference/matchers.html

For example this could maybe be:

```
EXPECT_THAT(build, Contains(std::make_pair("host", "foobar.xyz"));
```

Food for thought!

Reply:
Thanks for the suggestion, this makes sense. switched to `operator[]` which returns an empty string if the header is not set.

I also tried the `Contains` matcher but it did not compile because the value type of the request fields did not match. I also tried a helper which asserts that a `find` on the header field is not equal to `end` and then checks the value with `at`, but the subscript operator seems more concise to me.

Reply:
Agreed, thanks!